[Woo POS] [Cash & Receipts] Animations

iamgabrielma opened the pull request:
<!-- Remember about a good descriptive title. -->

Partially addresses #14808
https://github.com/woocommerce/woocommerce-ios/pull/14830 should be merged first.
<!-- Id number of the GitHub issue this PR addresses. -->
<!-- Part of: # -> use this if your PR is one of many related to one issue -->

## Description
This PR adds transitions/animations between:
- Transition between cash payment and order success screen

https://github.com/user-attachments/assets/f3466280-1776-4420-afab-55be35aedc6c

- Transition between order success screen and email receipt screen

https://github.com/user-attachments/assets/8a0f9c60-2ca2-4943-a0dd-eeb62b304c9e

## Testing information
- Set `.acceptCashForPointOfSale` flag to `true`
- Within a POS transaction, selecting to pay by cash, observe that the cash collection view appears under an animated transition
- Upon marking the order as complete (any or no amount), this also disappears with transition
- Tapping on "send receipt" appears/disappears with transition/animation as well.

Please note that this is an intermediate PR, so is enough to focus only in the animations for the review, other work like amount validation and UI fixes are dealt on separate PRs:
- https://github.com/woocommerce/woocommerce-ios/pull/14830
- https://github.com/woocommerce/woocommerce-ios/pull/14811

We do not have specifics from design regarding what sort of transitions should happen between views, happy to iterate further or change the transition direction as needed. For reference, here's a recording from Android: p1734619950830479?thread_ts=1734615162.822509&cid=C070SJRA8DP-slack-C070SJRA8DP

---
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

## Reviewer (or Author, in the case of optional code reviews):

Please make sure these conditions are met before approving the PR, or request changes if the PR needs improvement:

- [x] The PR is small and has a clear, single focus, or a valid explanation is provided in the description. If needed, please request to split it into smaller PRs.
- [x] Ensure Adequate Unit Test Coverage: The changes are reasonably covered by unit tests or an explanation is provided in the PR description.
- [x] Manual Testing: The author listed all the tests they ran, including smoke tests when needed (e.g., for refactorings). The reviewer confirmed that the PR works as expected on all devices (phone/tablet) and no regressions are added.

iamgabrielma committed:
animate receipt view from bottom

iamgabrielma committed:
add trailing animation for collectcashview

iamgabrielma commented on WooCommerce/Classes/POS/Presentation/PaymentButtons.swift:
```diff
@@ -65,7 +65,9 @@ private extension PaymentsActionButtons {
         let isEligible = await checkReceiptEligibility()
         if isEligible {
-            isShowingSendReceiptView = true
+            withAnimation {
```
I got different animation behaviour using `withAnimation` vs `.animation(value:)`. I'm not entirely sure why, but `withAnimation` seems to work better despite I like less how we have to call it. For context I've tested this with iOS 17

staskus replied:
What worked differently? IMHO, setting `isShowingSendReceiptView = true` without `withAnimation` is preferred since we can say exactly which views and how should apply the animation.

I suppose it can depend on where we put `.animation(value:)` in a view hierarchy but I haven't developed a good intuition yet on a precise way of working.

iamgabrielma replied:
I'm not sure how to explain it, is the same animation but looks more clunky 😅 ,definitely tricky to choose where to put it. I've removed this explicit call here: 4a392e33ea5b8d94301b0f46ca1ce511a77da5ce

staskus commented:
> Please note that this is an intermediate PR, so is enough to focus only in the animations for the review, other work like amount validation and UI fixes are dealt on separate PRs:
> https://github.com/woocommerce/woocommerce-ios/pull/14830

👍  One more UI update to do is to make sure the button doesn't get smaller when it's clicked. You can notice it shrinks in a slowed-down video.

staskus commented on WooCommerce/Classes/POS/Presentation/CardReaderConnection/UI States/Reader Messages/PointOfSaleCardPresentPaymentSuccessMessageView.swift:
```diff
@@ -13,6 +13,7 @@ struct PointOfSaleCardPresentPaymentSuccessMessageView: View {
     var body: some View {
         if isShowingSendReceiptView {
             POSSendReceiptView(isShowingSendReceiptView: $isShowingSendReceiptView)
+                .transition(.move(edge: .bottom))
```
One thing that I learned is to put both animated views within a single view hierarchy so `SwiftUI` would understand better how to animate transitions between them. For example, `VStack`. 

That could potentially be tried out is putting `POSSendReceiptView` and the rest of the success view within a single view hierarchy, for example, `VStack`, or `HStack`. You can see in the `PointOfSaleDashboardView`, how having an `HStack` allows `TotalsView` to slide in together while `CartView` slides out.

staskus replied:
You can also play around with `geometryGroup` (and `geometryGroupIfSupported` extension I added).

It signals `SwiftUI` to treat some collection of views as one, making them animate together.

iamgabrielma replied:
Wrapping them in a [common stack](6cc198dcef3d19fa48695f57b3dab178ebb108a6) seems to improve it a bit overall. I've also tried both `geometryGroup` and `matchedgeometry` but ends with a very odd results:

https://github.com/user-attachments/assets/62deb6a3-4872-4cfc-9908-60294059c9e0

https://github.com/user-attachments/assets/0b22a3fe-6db5-486e-83cb-2e38a93df59d

staskus reviewed (commented):
Thanks for the work!

> We do not have specifics from design regarding what sort of transitions should happen between views, happy to iterate further or change the transition direction as needed. For reference, here's a recording from Android: p1734619950830479?thread_ts=1734615162.822509&cid=C070SJRA8DP-slack-C070SJRA8DP

I think:
- Both Email Receipt and Cash Payment animations do not need to fade in / fade out
- It would be nice if Cash Payment would slide in while the other view moves towards the opposite edge synchronously. Similar to what is done with Cart -> Checkout transition, or developer on Android. 

I also noticed Android pushes both Cash and Email receipt views, while we push Cash payment but present Emails modally. I don't remember if it's deliberate or not.

-------

Technically, I know this animation work is complicated, and requires grasping the `SwiftUI` view update mechanims well. Let me know if you tried to do some of the things I wrote but It wouldn't work or created other issues. We could look for reasonable solutions.

iamgabrielma committed:
Merge branch 'trunk' into task/pos-cash-view-animations

iamgabrielma committed:
remove easyInOut anim from presenting cash view

iamgabrielma committed:
wrap in zstack to apply common animation

iamgabrielma committed:
remove duplicated matched geometry

iamgabrielma committed:
Keep button size consistent while loading

iamgabrielma committed:
remove explicit withAnimation call when show receipt view

iamgabrielma commented:
Thanks for the review @staskus ! I've tried several things, I'd say is more annoying to test anything else 😂 Ultimately what seems to look the best from several tries is a simple approach of:

- Wrap both components in a in ZStack
- Make the receipt transition `.leading`
- Make the rest transition `.trailing`
- Animate with `.default` and using `isShowingSendReceiptView` as value 

It still jumps a bit at the end, but on normal speed I don't think it's noticeable. Let me know what you think! Happy to look into more solutions or alternatives.

https://github.com/user-attachments/assets/5f9823d1-d5da-4bba-b9d7-815e4cb5369e

> One more UI update to do is to make sure the button doesn't get smaller when it's clicked. You can notice it shrinks in a slowed-down video.

Good catch, took me a few times to see. Addressed here: c4682dea9ff76b5df5fc71d58dd58d5648c22709

>  Let me know if you tried to do some of the things I wrote but It wouldn't work or created other issues. We could look for reasonable solutions.

Apart from the videos above using `geometryGroup` and `matchedgeometry` I've also tried others like `asymmetric` but most of them seem to conflict due having already `matchedgeometry` in the child components. Just as an example, those are not committed:

https://github.com/user-attachments/assets/5c8e454b-9109-47d7-b74b-1cd8d346c369

https://github.com/user-attachments/assets/b7c6e23e-e4de-45c4-b38e-42eef2e95d6f

staskus commented:
@iamgabrielma, are we planning to align these buttons? IMHO, they should be in the same place, use the same back icon, same text font, and size.

<img width="818" alt="image" src="https://github.com/user-attachments/assets/c15c0531-51eb-4883-bd4b-b7781decbfbc" />

staskus committed:
Merge branch 'trunk' into task/pos-cash-view-animations

staskus commented:
Low-priority. One more thing to log is to update this error message:

<img width="513" alt="image" src="https://github.com/user-attachments/assets/7184f65f-713a-4287-9d54-ba5fb59b2905" />

"To process this payment please connect the reader or choose cash."

staskus commented:
The floating button drops down after coming back from Cash view. It should just stay in a static position.

https://github.com/user-attachments/assets/e12c997f-cf21-4526-b3c8-97ead9571a51

staskus commented:
Email Receipt animates to the wrong direction, not forward but back. 

https://github.com/user-attachments/assets/6a37d40a-33cd-43b6-a9c2-e047a5a9e63e

staskus reviewed (commented):
>  I've tried several things, I'd say is more annoying to test anything else

Thanks!

I [myself used Injection](https://github.com/johnno1962/InjectionIII) to quickly iterate through animations. Enough to make a quick change, save, and retry opening views again. No need for a rebuild. 

The result still doesn't look solid to me. I know there are deficiencies in the previous screens with animations but I think it's a good opportunity to understand why something doesn't work the way we want.

I think the current Android implementation is a solid example.

Doing a full push transition from Checkout to Cash Payment view. Views stick together, separate labels or buttons are not moving on their own, the opacity doesn't change. 

Some other observations:
- [Misaligned back button](https://github.com/woocommerce/woocommerce-ios/pull/14854#issuecomment-2590365799)
- [Dropping floating button](https://github.com/woocommerce/woocommerce-ios/pull/14854#issuecomment-2590397933)
- [Email receipt should animate into the opposite direction](https://github.com/woocommerce/woocommerce-ios/pull/14854#issuecomment-2590466337)

iamgabrielma committed:
Render floating buttons only after the animation has finished

We render the POSFloatingControlView immediately after the payment state changes to .card, but this causes the button to jump around as the view is ongoing animations. We add a new check to render the component only when the animation has finished

iamgabrielma committed:
Merge branch 'task/pos-cash-view-animations' of https://github.com/woocommerce/woocommerce-ios into task/pos-cash-view-animations

iamgabrielma committed:
Update transition direction

Updates the transition direction so showing the receipt shows up as an unidirectional from, from success to receipt, and back if needed.

iamgabrielma committed:
update onscreen instructions

Updates the localized string ID as well so the new string does show appropiately

iamgabrielma committed:
Header alignment

We align the header for the cash view as the previous header on the navigation flow (cart)

iamgabrielma commented:
> I [myself used Injection](https://github.com/johnno1962/InjectionIII) to quickly iterate through animations. Enough to make a quick change, save, and retry opening views again. No need for a rebuild.

💯 

> The result still doesn't look solid to me. I know there are deficiencies in the previous screens with animations but I think it's a good opportunity to understand why something doesn't work the way we want.
> I think the current Android implementation is a solid example.
> Doing a full push transition from Checkout to Cash Payment view. Views stick together, separate labels or buttons are not moving on their own, the opacity doesn't change.

Agree that looks cleaner that way, I'm not entirely sure how I would go about pushing the view into the stack though, as currently all views in the TotalsView are presented based on state but not "pushed" into the stack? Or I'm miss-understanding how would work, which is also possible 🤔 

I find the current animations/transitions and views are getting more and more complicated and could be simplified or perhaps restructured into smaller bits, or not try to reuse some other bits, which adds complexity when trying to deal with multiple cases of how the view should be presented.

> Missaligned button

Sounds good. Updated on https://github.com/woocommerce/woocommerce-ios/pull/14854/commits/0e62a531e53977f801a2f67ad7245e22d2c9d73a

> Dropping floating button

Since this was rendered exclusively based on payment state, it was loaded before the animation for the rest of views was completed, which caused this jump. Fixed here: https://github.com/woocommerce/woocommerce-ios/pull/14854/commits/727f9ea710018f05695f32dc43b601df931c64f3

https://github.com/user-attachments/assets/04ceba97-e95b-43a1-905b-9a00f6766a16


> Animate in the opposite direction

I believe you mean to making them like an "unidirectional" flow, right? By switching the transitions here we accomplish that: https://github.com/woocommerce/woocommerce-ios/pull/14854/commits/8b79c4fa06f32a6916e24924490221711bcb62ae

https://github.com/user-attachments/assets/4d502eaf-39c2-416e-83e7-2a8ca84648fa

> Update this error message

Done! https://github.com/woocommerce/woocommerce-ios/pull/14854/commits/7beca97f79626c4a32396e0203a722d6388203ac

Let me know if this looks better, I'll check separately how we can make this easier via "pushing a view" without affecting the subviews within.

staskus reviewed (commented):
Thanks for the changes. Some observations:

> I find the current animations/transitions and views are getting more and more complicated and could be simplified or perhaps restructured into smaller bits, or not try to reuse some other bits, which adds complexity when trying to deal with multiple cases of how the view should be presented.

Yes, it's likely and expected that we'll need to iterate and improve our animation implementation while adding new animations.

> Missaligned button
> Sounds good. Updated on https://github.com/woocommerce/woocommerce-ios/commit/0e62a531e53977f801a2f67ad7245e22d2c9d73a

- [x]  It's still not aligned. I don't want pixel-perfectness to be a blocker but navigation is a critical thing and merchants will be tapping an expected part of the screen to go back. These misalignments could be of annoyance. 

<img width="1733" alt="not aligned" src="https://github.com/user-attachments/assets/608a06d9-d9a6-4921-a375-a80493759d2c" />

- [x] The same goes with Email Receipts which now also look very differently from what we use in < Cart and < Cash Payment.

![Email Receipt Button not aligned](https://github.com/user-attachments/assets/94cad66f-91f3-4037-b47d-14092dd78776)

- [x] We may also need to reconsider the titles of the back buttons. While "< Cart" signifies that we're coming back to the cart, "< Cash Payment" and "< Email Receipt" signifies our current screen. But it's outside the scope of this task for sure.

> Since this was rendered exclusively based on payment state, it was loaded before the animation for the rest of views was completed, which caused this jump. Fixed here: https://github.com/woocommerce/woocommerce-ios/commit/727f9ea710018f05695f32dc43b601df931c64f3

- [ ]  The issue with the floating button happens because it's tied to keyboard presentation, or likely some safe area insets. The floating button shouldn't rise at all when the keyboard rises to avoid it jumping around.

https://github.com/user-attachments/assets/be933b2b-5f8e-4f42-9172-d9c92fa3f454

- [ ] (Slightly related to the previous point), we could also try hiding the keyboard earlier before starting the animation. I'm not sure if that would work but [likely would mean force-hiding the keyboard when the back is tapped.](https://www.hackingwithswift.com/quick-start/swiftui/how-to-dismiss-the-keyboard-for-a-textfield)

<img width="500" alt="image" src="https://github.com/user-attachments/assets/7227708c-d8f9-4dc8-955a-fa63aaa42688" />

- [x] Regarding animations. It would be great to avoid the sudden disappearance of a view while it's getting pushed out. We can see half of the "Payment Successful" still visible before the view appears.

https://github.com/user-attachments/assets/ff9a1ab7-aade-44f0-8d59-00e19e69b8ab

---------

Let me know which things you're addressing on this task and the end result you're aiming for. I judge many of these things subjectively not based on concrete goals.

iamgabrielma committed:
Adjust cash view navigation top header padding

iamgabrielma committed:
Adapt send receipt view style for consistency with rest of cash-receipt navigation

iamgabrielma committed:
Address sudden disappearance of view rather than animated

iamgabrielma commented:
Thanks for the review again @staskus , apologies for the back and forth 🙇 

> Header alignment

I believe this is fully fixed now, one of the views was missing a top padding, and the receipts view being consistent I've missed entirely, so I used the same configuration than for the other views for consistency. In the video below I just open the text editor aligned to the bottom of the text so is easier to record alignment across different screens:

https://github.com/user-attachments/assets/a6e38932-824d-4b1b-9ab5-1aa2f165527c


> Sudden view disappearance 

Using `.asymetric` transitions for insert/remove seems to resolve the issue I would expect to use different values on insertion and removal. If I do so, this causes a regression on how the views appear and no longer show on an "uni-directional" manner. I'm not entirely sure why, but adding the same direction to both insertion and removal works correctly.

https://github.com/user-attachments/assets/dd0a04aa-f770-40d1-8011-f7d438c931fc

> Floating button/keyboard

I haven't addressed these, I can log it separately 

> We may also need to reconsider the titles of the back buttons. While "< Cart" signifies that we're coming back to the cart, "< Cash Payment" and "< Email Receipt" signifies our current screen. But it's outside the scope of this task for sure.

Agree, is a bit odd. I pinged design in the testing P2 regarding this issue.

staskus commented on WooCommerce/Classes/POS/Presentation/PointOfSaleDashboardView.swift:
```diff
@@ -71,6 +72,13 @@ struct PointOfSaleDashboardView: View {
+                    isAnimating = false
+                }
+            }
+        }
```
This solves the problem of the floating button going up when the keyboard appears/disappears issue. 

Annoyingly, it has to be applied to the parent, rather than child view but none of our views need to adapt to keyboard safe area. 

This allows to remove `isAnimating` check.

```suggestion
        .ignoresSafeArea(.keyboard)
```

staskus commented on WooCommerce/Classes/POS/Presentation/PointOfSaleDashboardView.swift:
```diff
@@ -36,7 +37,7 @@ struct PointOfSaleDashboardView: View {
                     .trackSize(size: $floatingSize)
                     .accessibilitySortPriority(1)
-                    .renderedIf(posModel.itemsViewState.containerState != .loading)
+                    .renderedIf(posModel.itemsViewState.containerState != .loading && !isAnimating)
```
```suggestion
                    .renderedIf(posModel.itemsViewState.containerState != .loading)
```

staskus commented on WooCommerce/Classes/POS/Presentation/PointOfSaleDashboardView.swift:
```diff
@@ -7,6 +7,7 @@ struct PointOfSaleDashboardView: View {
     @State private var showSupport: Bool = false
 
     @State private var floatingSize: CGSize = .zero
+    @State private var isAnimating: Bool = false
```
```suggestion
```

staskus reviewed (approved):
Thanks for the patience 👍 

The animations look good! 🚀 

#### Floating button issue
I left one suggestion to deal with it so it would ignore keyboard altogether and won't be going one when we're dismissing the view.

#### Delayed keyboard dismissal issue

I think it's related to the fact that we're using custom transitions, so the view disappears later than we want to dismiss the keyboard.

I implemented the change while testing, I will make a separate PR for this (you can check the video):

https://github.com/woocommerce/woocommerce-ios/pull/14902

#### Back buttons

Thanks for fixing the alignment!

The POS back button should share a single component to match each other. We have different spacing between the arrow and the text and different sizes of the arrow.

<img width="184" alt="Email Receipt padding" src="https://github.com/user-attachments/assets/d19e9831-c8bf-442e-9301-1cea9dafea29" />
<img width="174" alt="Cash Payment padding" src="https://github.com/user-attachments/assets/a153f72d-0d68-46aa-824e-3ee839ed0840" />
<img width="98" alt="Cart button padding" src="https://github.com/user-attachments/assets/5bc42e50-37cd-41b2-8c20-e97c2bf2577b" />

Created a task:
- https://github.com/woocommerce/woocommerce-ios/issues/14903

iamgabrielma committed:
Apply suggestions from code review

By ignoring the keyboard safe area we can get rid of isAnimating logic for the floating button appearance-disappearance.

Co-authored-by: Povilas Staskus <povilas.staskus@automattic.com>

iamgabrielma commented:
> I left one suggestion to deal with it so it would ignore keyboard altogether and won't be going one when we're dismissing the view.

That's interesting, I wouldn't have thought to resolve it like that. 🤔 thanks! I confirmed and committed your suggestion
